### PR TITLE
ignore-deleting-ebs-volume-from-diferent-cluster

### PR DESCRIPTION
**Pull request by calvix**
towards https://github.com/giantswarm/giantswarm/issues/14020

during a deletion, volumes attached to an instance that belongs to a different `cluster-id` will be ignored

## Checklist

- [x] Update changelog in CHANGELOG.md.

### CODE REVIEW

File: service/internal/ebs/ebs.go [hunk 1]
```diff
@@ -208,25 +208,60 @@ func (e *EBS) ListVolumes(cr infrastructurev1alpha2.AWSCluster, filterFuncs ...f
...
-		}
+		if !ignoreVolume {
+			volume := Volume{
+				VolumeID:    *v.VolumeId,
+				Attachments: attachments,
+			}
 
-		volumes = append(volumes, volume)
+			volumes = append(volumes, volume)
+		}
```
Comment (calvix) [inline service/internal/ebs/ebs.go:253]: I wonder whats better to read that or this
```
if ignoreVolume {
  continue
}
```

File: service/internal/ebs/ebs.go [hunk 2]
```diff
@@ -208,25 +208,60 @@ func (e *EBS) ListVolumes(cr infrastructurev1alpha2.AWSCluster, filterFuncs ...f
...
+					}
+
+					o, err := e.client.DescribeInstances(i)
+					if err != nil {
+						return nil, microerror.Mask(err)
+					}
+
+					// if the instance does not have the proper cluster-id tag than ignore the volume
+					if len(o.Reservations) > 0 && len(o.Reservations[0].Instances) > 0 &&
+						!instanceHasTag(o.Reservations[0].Instances[0], key.TagCluster, cr.Labels[key.TagCluster]) {
```
Comment (xh3b4sd) [inline service/internal/ebs/ebs.go:229]: Can we make this condition a little bit shorter and more readable? 

```go
hasInstances := len(o.Reservations) > 0 && len(o.Reservations[0].Instances) > 0
hasTags := instanceHasTag(o.Reservations[0].Instances[0], key.TagCluster, cr.Labels[key.TagCluster])

if hasInstances && !hasTags {
    ...
}
```

Comment (xh3b4sd) [reply service/internal/ebs/ebs.go:229]: I am also wondering if we need to pay attention to `o.Reservations[0].Instances[1]` and beyond. Are there more instances necessary to check?

File: service/internal/ebs/ebs.go (hunk 1 above)
Comment (xh3b4sd) [reply service/internal/ebs/ebs.go:253]: An early continue makes sense to me.

File: service/internal/ebs/ebs.go (hunk 2 above)
Comment (calvix) [reply service/internal/ebs/ebs.go:229]: EBS volume can be only attached to a single instance so We don't need to care, also the request to desciribe instances is via `instance-id` which is unique across AWS so result of that request must be a single instance

Comment (calvix) [reply service/internal/ebs/ebs.go]: > hasInstances := len(o.Reservations) > 0 && len(o.Reservations[0].Instances) > 0
hasTags := instanceHasTag(o.Reservations[0].Instances[0], key.TagCluster, cr.Labels[key.TagCluster])
if hasInstances && !hasTags {
    ...
}

this does not work as if there are no instances the `hasTags` expression will panics

File: service/internal/ebs/ebs.go [hunk 3]
```diff
@@ -208,25 +208,60 @@ func (e *EBS) ListVolumes(cr infrastructurev1alpha2.AWSCluster, filterFuncs ...f
...
 
-		volumes = append(volumes, volume)
+			volumes = append(volumes, volume)
+		}
 	}
 
 	return volumes, nil
 }
+
+func instanceHasTag(instance *ec2.Instance, tagKey string, tagValue string) bool {
```
[the remark this answers is not in the document]
Comment (calvix) [reply service/internal/ebs/ebs.go]: sorted

**Review by xh3b4sd — APPROVED** [review]

**Pull request merged**